### DISCUSSION TITLE
edge_label_index sometimes returns edges labeled as positive even if those edges don't exist in the sampled subgraph

### DISCUSSION THREAD

**Discussion by jlimsf** [pyg-team/pytorch_geometric#6518] [Q&A] [answered]
i am trying to find the explanantion behind this. i found the corresponding documentation (i think), but am still a little confused: https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/loader/link_neighbor_loader.html

print(x)
    print(x['edge_index'])
    print(x['edge_label_index'])
    print(x['edge_label'])

```
Data(x=[16, 12], edge_index=[2, 9], input_id=[2], edge_label_index=[2, 4], edge_label=[4]).        #x
tensor([[ 0,  8,  9, 10, 11, 12, 13, 14, 15],         #x['edge_index'] #
        [ 1,  1,  2,  2,  2,  2,  5,  5,  7]])
tensor([[4, 0, 6, 7],           #x['edge_label_index'] 
        [2, 1, 5, 3]])
tensor([1., 1., 0., 0.]).          # x['edge_label'] 
```

in this example, edge `[4,2]` is labeled as positive but is not present in `x['edge_index']`. what's the reasoning behind this? thanks!

edit: i see that some documentation on the negative sampling indicates that there will be false negatives (https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.LinkLoader) since we are doing some approximations. Does that mean there will be false positives for the positive edges as well?

**Comment by EdisonLeeeee** [comment]
> Does that mean there will be false positives as well?

Definitely not. Actually, the returned `edge_label_index` in each batch was relabeled to hold consecutive indices starting from zero.

**Reply by jlimsf** [reply]
thanks for the reply. i'm not too sure what you mean here. what is the exact explanation for `[4,2]` being labeled as positive, then. `[4,2]` is not in `edge_index`

**Reply by EdisonLeeeee** [reply]
Sorry for not making it clear. In your case, `[4,2]` is mapped from an edge from the original graph (e.g.,`[9,2]`) `edge_index`. For example:
```python
tensor([[4, 0, 6, 7],           #relabeled x['edge_label_index'] 
        [2, 1, 5, 3]])
```
in the original graph might be
```python
tensor([[9, 0, 13, 15],           #original x['edge_label_index'] 
        [2, 1, 5, 7]])
```

**Reply by jlimsf** [reply]
thanks for the reply. 

interesting -so are there three coordinate systems we are working with? 

- global: coordinate space of the input graph
- subgraph/local1: the batch subgraph (`edge_index`)
- pos and neg examples/local2: the samples from within the batch we are training with (`edge_label_index`)

**Reply by jlimsf** [reply]
also, i thought in this example that `[6,5] / [13,5]` and `[7,3]/[15,7]` were not supposed to be valid edges?

**Comment by jlimsf** [comment]
wanted to add some additional context here.. 

Here is my code: 


```
edge_idxs = torch.tensor([
    [0, 1, 2, 2, 3, 3, 4],
    [2, 2, 3, 4, 5, 6, 7],
])
node_feats = torch.tensor([0,1,2,3,4,5,6,7])
sample = Data(x=node_feats, edge_index=edge_idxs)
sample_loader = LinkNeighborLoader(sample, batch_size=1, shuffle=True,
                            neg_sampling='binary', num_neighbors=[1,],
                            num_workers=0,)
for x in sample_loader:

    edge_index_transposed = x['edge_index'].T.numpy()
    edge_label_index_transposed = x['edge_label_index'].T.numpy()
    positive_edges_idxs = (x['edge_label'] == True).tolist()
    positive_edges = edge_label_index_transposed[positive_edges_idxs]
    
    for pos in positive_edges:
        try:
            assert(pos.tolist() in edge_index_transposed.tolist())
        except :
            print(x['edge_index'])
            print(x['edge_label_index'])
            print(x['edge_label'])
            raise
```
an example print/AssertionError:

```
tensor([[4, 1, 5],
        [1, 2, 3]])
tensor([[0, 3],
        [1, 2]])
tensor([1., 0.])

```

**Reply by jlimsf** [reply]
edge `[0,1]` is labeled as a positive edge but is nowhere in the sampled graph or the original graph.

**Comment by jlimsf** [comment, marked as answer]
@EdisonLeeeee i may be misinterpreting your response, but i found the answer to my confusion. there are no false positives. basically, a subgraph is created at every batch and then the positive nodes are valid edges within this subgraph. since i have `num_neighbors = [1[`, i will run into cases where the positive edges are not `x['edge_index']`. if i do `num_neighbors=[-1]`, i see the positive edge show up. furthermore, i created some test code to test certain properties of the subgraph sampled at every batch which made things clear for me. all positive edges are in this batch-level subgraph.
